### PR TITLE
set template variable when clicking self dashboard link

### LINKED ISSUE #9259: Table panel: Update current dashboard template variable via cell links
- Grafana v4.6.0

To reproduce:

1. Set a table-column to `Render value as link`

1. Create a link that points to the current dashboard but with changed variables (e.g. `https://example.com/dashboard/db/overview?var-server=$__cell`)

1. Try to click the link. The address bar changes but the page does not reload.

### PR DESCRIPTION
**Pull request by mtanda**
Another approach, https://github.com/grafana/grafana/pull/16067 .
Fix https://github.com/grafana/grafana/issues/9259 and https://github.com/grafana/grafana/issues/8249

### CODE REVIEW

**Comment by mtanda** [comment]
Test failure message. Not sure why the test doesn't pass...

```
>> Summary of all failing tests
>> FAIL public/app/plugins/datasource/stackdriver/specs/datasource.test.ts
>>   ● StackdriverDataSource › When performing query › and no time series data is returned › should return a list of datapoints
>>
>>     TypeError: Object.values is not a function
>>
>>       113 |     const data = await this.getTimeSeries(options);
>>       114 |     if (data.results) {
>>     > 115 |       Object['values'](data.results).forEach(queryRes => {
>>           |                       ^
>>       116 |         if (!queryRes.series) {
>>       117 |           return;
>>       118 |         }
>>
>>       at StackdriverDatasource.<anonymous> (public/app/plugins/datasource/stackdriver/datasource.ts:115:23)
>>       at step (node_modules/tslib/tslib.js:133:27)
>>       at Object.next (node_modules/tslib/tslib.js:114:57)
>>       at fulfilled (node_modules/tslib/tslib.js:104:62)
>>
>>
>> Test Suites: 1 failed, 222 passed, 223 total
>> Tests:       1 failed, 1 skipped, 1896 passed, 1898 total
>> Snapshots:   103 passed, 103 total
>> Time:        80.852s
>> Ran all test suites.
```

**Comment by chicco785** [comment]
@mtanda so basically with this PR, if i have a link setting a variable in a panel, I am able to update an other panel in the same dashboard (assuming the other panel outcome depends on that variable)?

**Comment by mtanda** [comment]
I didn't test it, but it should be updated.

**Comment by michal-rumanek** [comment]
@mtanda what about your fix, do you want to merge it?

**Comment by mtanda** [comment]
@michal-rumanek Yes. If I'm lucky, core team will review this. If not, I wait patiently.

**Comment by mtanda** [comment]
@torkelo ping, if there is no chance to merge this change, please notify me :-)

**Comment by mtanda** [comment]
@marefr Would you look this PR?

**Comment by Stono** [comment]
Can anyone look at this? It fixes a really annoying issue!

**Comment by mtanda** [comment]
I'm waiting for review. remove stale label.

**Comment by mtanda** [comment]
I'm waiting for review. remove stale label again.

**Comment by mtanda** [comment]
ping

**Comment by mtanda** [comment]
ping

**Comment by mtanda** [comment]
ping

File: public/app/features/templating/variable_srv.ts [hunk 1]
```diff
@@ -303,6 +311,42 @@ export class VariableSrv {
...
+    if (path.indexOf('?') >= 0) {
+      path = path.slice(0, this.originalUrl.indexOf('?'));
+    }
+    if (state.location.path === path) {
+      if (state.location.url === this.originalUrl) {
+        return this.$q.when({});
+      }
+
+      this.originalUrl = this.$location.url();
+      const queryParams = this.$location.search();
```
Comment (ryantxu) [inline public/app/features/templating/variable_srv.ts]: When updating via `getLocationSrv().update( { ... } )` this works when I change the above line to:
```
const queryParams = state.location.query; //this.$location.search();
```

@mtanda -- for your test cases does that work?  or does it need to be $location.search()?

**Comment by ryantxu** [comment]
@mtanda -- this looks great, and I think very close.  You can see my similar struggles in: #21480

**Comment by ryantxu** [comment]
@mtanda  -- check #21480 -- this is an alternative approach that listens to the redux store and then fires an event

**Comment by mtanda** [comment]
@ryantxu Your approach is much better than my approach!
Thanks! I hope the fix is included in next Grafana release :-)

**Pull request closed without merging**